lib/cdo/i18n_backend.rb

```diff
@@ -164,12 +164,104 @@ def translate(locale, key, options = ::I18n::EMPTY_HASH)
       end
     end
 
-    class SimpleBackend < ::I18n::Backend::Simple
+    module Plugins
       include SmartTranslate
       include MarkdownTranslate
       include SafeInterpolation
       include I18nStringUrlTrackerPlugin
     end
+
+    class SimpleBackend < ::I18n::Backend::Simple
+      include Plugins
+    end
+
+    # The class extends the `LazyLoadable` backend from the i18n gem to support CDO i18n files:
+    # - Modified the `LocaleExtractor` class to support filenames that include both a prefix and the locale (e.g., "courses.en-US.yml").
+    # - Modified the `filenames_for_current_locale` method to load files for the current locale and its fallbacks, (e.g, "en" for "en-US" and vice versa)
+    # - Modified the `eager_load!` method to ignore eager loading when lazy loading is enabled.
+    # - Modified the `assert_file_named_correctly!` method to check if a filename corresponds to the locale it loaded.
+    # - Implemented a mechanism to avoid reloading already loaded translation files.
+    # https://github.com/ruby-i18n/i18n/blob/v1.12.0/lib/i18n/backend/lazy_loadable.rb
+    class LazyLoadableBackend < ::I18n::Backend::LazyLoadable
+      include Plugins
+
+      LOCALES_MAPPING = YAML.load_file(CDO.dir('dashboard/config/locales.yml')).each_with_object({}) do |(k, v), locales|
+        locales[k.to_sym] = v.to_sym if v.is_a?(String)
+      end.freeze
+
+      # The original method has been modified to works on i18n files named with a prefix and the locale,
+      # like "common.en.yml" or "common.en-US.json" and not only on files named with the locale, like "en-US.json".
+      # https://github.com/ruby-i18n/i18n/blob/v1.12.0/lib/i18n/backend/lazy_loadable.rb#L55-L63
+      class ::I18n::Backend::LocaleExtractor
+        def self.locale_from_path(path)
+          # Extracts the locale from the path, like "en", "en-US", "haw", "haw-HI", etc.
+          path[/\b([a-z]{2,3}(?:-[A-Z]{2})?)\b\.\w+$/, 1]&.to_sym
+        end
+      end
+
+      def load_translations(*files)
+        loaded_files.merge(files.flatten) if lazy_load?
+        super
+      end
+
+      def reload!
+        @loaded_files = nil if lazy_load?
+        super
+      end
+
+      def eager_load!
+        # Ignores eager loading instead of raising an error if lazy loading is enabled
+        super unless lazy_load?
+      end
+
+      protected
+
+      def loaded_files
+        @loaded_files ||= Set.new
+      end
+
+      # Get the set of valid locales for a given base locale code. For example,
+      # for a base of `:pt`, we expect valid locales `:pt`, `:"pt-BR"`, `:"en-US"`, and `en`
+      def valid_locales_for(base_locale)
+        valid_locales = Set.new(::I18n.fallbacks[base_locale])
+        valid_locales << base_locale
+        valid_locales << LOCALES_MAPPING[base_locale] # en: :'en-US'
+        valid_locales.compact_blank
+      end
```

Comment:
Is it worth memoizing this method? I don't think we expect the results to ever change for a given locale, and with the changes to `assert_file_named_correctly` it seems like it's going to get called pretty frequently.

Reply:
Do you mean memoizing in a hash per locale?

Reply:
Yep! Just something simple. Not super important, but could be a nice optimization